### PR TITLE
Add parser changes for included record parameter

### LINKED ISSUE #26123: Parser changes for included record parameter feature
**Description:**
Please refer https://github.com/ballerina-platform/ballerina-spec/issues/578#issue-671556255

### PR DESCRIPTION
**Pull request by chiranSachintha**
This PR addressed review suggestions mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/26882

Fixes #26123 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/TreeModifier.java [hunk 1]
```diff
@@ -725,14 +725,28 @@ public RequiredParameterNode transform(
...
+            IncludedRecordParameterNode includedRecordParameterNode) {
+        NodeList<AnnotationNode> annotations =
+                modifyNodeList(includedRecordParameterNode.annotations());
+        Token asteriskToken =
+                modifyToken(includedRecordParameterNode.asteriskToken());
+        Node typeName =
+                modifyNode(includedRecordParameterNode.typeName());
+        Token paramName =
+                modifyToken(includedRecordParameterNode.paramName().orElse(null));
+        return includedRecordParameterNode.modify(
```
Comment (dulvinw) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/TreeModifier.java:749]: Looks like we can improve the formatting?

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java [hunk 2]
```diff
@@ -381,6 +382,31 @@ public RequiredParameterNode transform(RequiredParameterNode requiredParameterNo
...
 
+    @Override
+    public IncludedRecordParameterNode transform(IncludedRecordParameterNode includedRecordParameterNode) {
+        NodeList<AnnotationNode> annotations = formatNodeList(includedRecordParameterNode.annotations(), 1, 0, 1, 0);
+        Node typeName;
+        Token asterisk = formatToken(includedRecordParameterNode.asteriskToken(), 0, 0);
+        if (includedRecordParameterNode.paramName().isPresent()) {
+            typeName = formatNode(includedRecordParameterNode.typeName(), 1, 0);
+            Token paramName =
+                    formatToken(includedRecordParameterNode.paramName().orElse(null), env.trailingWS, env.trailingNL);
```
Comment (dulvinw) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:393]: Here as well. Arguments are usually aligned AFAIK

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/TreeModifier.java (hunk 1 above)
Comment (chiranSachintha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/TreeModifier.java:749]: This is an auto-generated code segment and this formatting issue is with other places also. (In this file)

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java (hunk 2 above)
Comment (chiranSachintha) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java]: Fixed.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 3]
```diff
@@ -310,7 +310,7 @@
             { ParserRuleContext.CLOSE_PARENTHESIS, ParserRuleContext.REQUIRED_PARAM };
 
     private static final ParserRuleContext[] PARAMETER_START =
-            { ParserRuleContext.ANNOTATIONS, ParserRuleContext.TYPE_DESC_IN_PARAM };
+            { ParserRuleContext.ANNOTATIONS, ParserRuleContext.TYPE_DESC_IN_PARAM, ParserRuleContext.ASTERISK};
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java]: Fixed.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -1956,13 +1956,13 @@ private STNode parseParameter(SyntaxKind prevParamKind, boolean isParamNameOptio
             case AT_TOKEN:
                 annots = parseOptionalAnnotations();
                 if (peek().kind == SyntaxKind.ASTERISK_TOKEN) {
-                    inclusionSymbol = parseAsteriskToken();
+                    inclusionSymbol = consume();
                 } else {
                     inclusionSymbol = STNodeFactory.createEmptyNode();
                 }
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: Changed.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -1950,27 +1992,16 @@ private STNode parseParameterRhs(SyntaxKind tokenKind) {
...
         switch (nextToken.kind) {
             case AT_TOKEN:
                 annots = parseOptionalAnnotations();
-                if (peek().kind == SyntaxKind.ASTERISK_TOKEN) {
-                    inclusionSymbol = parseAsteriskToken();
-                } else {
-                    inclusionSymbol = STNodeFactory.createEmptyNode();
-                }
-                break;
+                return parseParameter(annots, prevParamKind, isParamNameOptional);
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:1999]: Can replace with `break`

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 6]
```diff
@@ -2497,13 +2505,16 @@ protected ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLo
...
+                switch (parentCtx) {
+                    case ARRAY_TYPE_DESCRIPTOR:
+                        return ParserRuleContext.CLOSE_BRACKET;
+                    case XML_ATOMIC_NAME_PATTERN:
+                        endContext();
+                        return ParserRuleContext.XML_NAME_PATTERN_RHS;
+                    case REQUIRED_PARAM:
+                        return ParserRuleContext.TYPE_DESC_IN_PARAM;
+                    default:
+                        return ParserRuleContext.TYPE_REFERENCE_IN_TYPE_INCLUSION;                }
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java:2517]:
```suggestion
                        return ParserRuleContext.TYPE_REFERENCE_IN_TYPE_INCLUSION;               
                         }
```

Comment (chiranSachintha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java]: Fixed.

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java [hunk 7]
```diff
@@ -161,8 +161,8 @@ public void test(String fileName, String path) throws IOException {
                 "annot_decl_source_02.bal", "class_def_source_11.bal", "object_type_def_source_26.bal",
                 "ambiguity_source_32.bal", "service_object_constructor_source_01.bal",
                 "service_object_constructor_source_02.bal", "object_constructor_source_08.bal",
-                "annotations_source_01.bal", "do_stmt_source_08.bal", "func_params_source_10.bal",
-                "func_params_source_11.bal", "func_params_source_13.bal");
+                "annotations_source_01.bal", "do_stmt_source_08.bal", "func_params_source_14.bal",
+                "func_params_source_15.bal", "func_params_source_13.bal", "func_params_source_11.bal");
```
Comment (lochana-chathura) [inline misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java:165]: we need to run the formatter for negative tests as well. Shall we not not skip  func_params_source_13., func_params_source_14 and func_params_source_15

func_params_source_11 is fine as the formatter currently does not loop the invalid tokens in the minutiae.

**Review by sanjana — CHANGES_REQUESTED** [review]

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java (hunk 7 above)
Comment (sanjana) [inline misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java:165]: As an example, when formatting the following code sample with syntax errors, 

```
function foo(int a, int b, *person) {
}
```

the appropriate formatting would be,

```
function foo(int a, int b, *person ) {
}
```
Please note the _space_ after **`*person`**.

Here even if multiple tokens are expected after `*person` by the parser, still only a single space should be present. 

So, to enable such negative test cases, the relevant parser test source files should be reformatted accordingly.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java [hunk 8]
```diff
@@ -381,6 +382,31 @@ public RequiredParameterNode transform(RequiredParameterNode requiredParameterNo
...
+                    .withParamName(paramName)
+                    .apply();
+        } else {
+            typeName = formatNode(includedRecordParameterNode.typeName(), env.trailingWS, env.trailingNL);
+            return includedRecordParameterNode.modify()
+                    .withAsteriskToken(asterisk)
+                    .withAnnotations(annotations)
+                    .withTypeName(typeName)
+                    .apply();
+        }
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:407]: Not sure if early exit pattern is very useful here. Usually we exit early when there's a guard block or something.
In requiredParameterNode we have used a if-else block for the same scenario. So better stick to the same pattern?

[the remark this answers is not in the document]
Comment (rdulmina) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:407]: @lochana-chathura I think what @sanjana saying is we can simplify this code by removing `else` block as suggested. So +1 for that

[the remark this answers is not in the document]
Comment (chiranSachintha) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java]: Changed.

File: misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java (hunk 7 above)
Comment (chiranSachintha) [reply misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java]: Change as mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/27662#discussion_r553565834

Comment (chiranSachintha) [reply misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java]: Fixed.

**Review by lochana-chathura — APPROVED** [review]
LGTM

**Review by rdulmina — APPROVED** [review]

**Review by sanjana — APPROVED** [review]
LGTM

**Pull request merged**